resharper/resharper-unity/src/HlslSupport/UnityGameEngineDirectiveResolver.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using JetBrains.Annotations;
+using JetBrains.Application;
+using JetBrains.Collections.Viewable;
+using JetBrains.ProjectModel;
+using JetBrains.ReSharper.Plugins.Unity.ProjectModel;
+using JetBrains.ReSharper.Psi.Cpp.Parsing.Preprocessor;
+using JetBrains.ReSharper.Psi.Cpp.Util;
+using JetBrains.Util;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace JetBrains.ReSharper.Plugins.Unity.HlslSupport
+{
+    [SolutionComponent]
+    public class UnityGameEngineDirectiveResolver : IGameEngineIncludeDirectiveResolver
+    {
+        private readonly ISolution mySolution;
+        private readonly UnitySolutionTracker mySolutionTracker;
+        private readonly UnityReferencesTracker myReferencesTracker;
+        private readonly ILogger myLogger;
+
+        public UnityGameEngineDirectiveResolver(ISolution solution, UnitySolutionTracker solutionTracker, UnityReferencesTracker referencesTracker, ILogger logger)
+        {
+            mySolution = solution;
+            mySolutionTracker = solutionTracker;
+            myReferencesTracker = referencesTracker;
+            myLogger = logger;
+        }
+        
+        public bool IsApplicable(CppInclusionContext context, string path)
+        {
+              return path.StartsWith("Packages/") &&
+                   (mySolutionTracker.IsUnityProject.HasTrueValue() || myReferencesTracker.HasUnityReference.HasTrueValue());
```

Comment:
Do we need to do anything to support includes relative to `Assets/` here?

Reply:
No, when shader is used from Assets folder, it uses real file system path related to solution folder (https://youtrack.jetbrains.com/issue/RIDER-47412)